### PR TITLE
Page Table Walker implementation 

### PR DESCRIPTION
**Pull request by vishalgupta97**
This PR adds a five level page table walker with four paging structure caches (PSC) which stores intermediate translations. CR3 register stores the physical address of the first level (L5). The four PSCs are:-
PSCL5 : L5->L4
PSCL4: L5->L3
PSCL3: L5->L2
PSCL2: L5->L1

All PSCs are searched simultaneously and if hit in multiple caches, then this order will be followed: PSCL2 > PSCL3 > PSCL4 > PSCL5. 

A STLB miss will trigger a page table walk. A hit in PSC will determine from which level to start the walk. A request will be sent to L2 cache and PTW will wait for the response before sending the request for the next level. Once all levels are done, PSCs will be updated and response is sent to STLB. Currently, a single page walk is done at a time, but it can be extended to support multiple page walks simultaneously.

Page table pages are allocated in physical memory using the virtual memory va_to_pa function. The virtual address for the page table starts with 0xf000000f00000000 and is allocated sequentially a page at a time. This virtual address does not match with the virtual addresses present in SPEC 2017 and IPC traces.

Whenever there is a page fault, it is serviced in the same cycle and no penalty is added or any request sent to cache hierarchy. write_translation_page function is not implemented as I not sure how to model 8 cache blocks write per page level whenever there is a page fault. The current implementation is similar to an OS which uses non cacheable writes to write page table pages. 
  
Known issue: Prefetch request in virtual address space directly get the translation from va_to_pa function. Translation for L1I prefetch request takes zero cycles whereas for L1D a fixed latency is added currently. How to integrate this with TLB and PTW ?

### CODE REVIEW

**Comment by ngober** [comment]
Now I see why Seth was so active with my PRs yesterday. :smile:

I'm self-requesting a review on this. I like what this does on first glance, but I want to write a more fine-grained review of it. I expect that this may fill a need we've had for a while. Thank you for your willingness to submit this PR.

**Comment by sethpugsley** [comment]
That was just a coincidence.  I was unsure of the timing of this PR.

There's a lot here, and I'm trying to absorb it all, but one thing that jumps out to me is that you aren't using the get_pte_pa() function from VirtualMemory to map translation pages.  Instead, you're using the regular va_to_pa() function to map both data and translation pages, and you're assigning a new range of dummy VAs to your page table.  I intended the get_pte_pa() function to simplify this and get rid of the need for dummy VAs for the page table.

Can you explain the reasoning here?  I'm especially interested to know if there's some flaw in the way I imagined the get_pte_pa() function working that made it so you needed to re-do that part.  Is it just that you'd already had all this work in place when I introduced the VirtualMemory class?  If that's the case, then could you please convert your code to use get_pte_pa()?

**Comment by sethpugsley** [comment]
I forgot to say thanks for contributing this!  This was a lot of work that should significantly improve our simulation accuracy.

I guess the previous PR merging created some new conflicts that need to be addressed.  Also, here are some results I obtained just before the automatic merging got broken that I have some questions about.

Here's the command I used:
`/bin/champsim -warmup_instructions 50000000 -simulation_instructions 50000000 -traces ../../dpc3/spec2017_traces/600.perlbench_s-570B.champsimtrace.xz`

and here's the output from the develop branch running that code:
`CPU 0 cumulative IPC: 1.31878 instructions: 50000003 cycles: 37913751
L1D TOTAL     ACCESS:   17058774  HIT:   16897805  MISS:     160969
L1D LOAD      ACCESS:    7297545  HIT:    7154084  MISS:     143461
L1D RFO       ACCESS:    9761229  HIT:    9743721  MISS:      17508
L1D PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
L1D WRITEBACK ACCESS:          0  HIT:          0  MISS:          0
L1D PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
L1D AVERAGE MISS LATENCY: 66.8092 cycles
L1I TOTAL     ACCESS:    5250255  HIT:    4992032  MISS:     258223
L1I LOAD      ACCESS:    5250255  HIT:    4992032  MISS:     258223
L1I RFO       ACCESS:          0  HIT:          0  MISS:          0
L1I PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
L1I WRITEBACK ACCESS:          0  HIT:          0  MISS:          0
L1I PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
L1I AVERAGE MISS LATENCY: 8.95662 cycles
L2C TOTAL     ACCESS:     467349  HIT:     389750  MISS:      77599
L2C LOAD      ACCESS:     401684  HIT:     329160  MISS:      72524
L2C RFO       ACCESS:      17508  HIT:      12655  MISS:       4853
L2C PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
L2C WRITEBACK ACCESS:      48157  HIT:      47935  MISS:        222
L2C PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
L2C AVERAGE MISS LATENCY: 133.776 cycles
LLC TOTAL     ACCESS:      93861  HIT:      35432  MISS:      58429
LLC LOAD      ACCESS:      72524  HIT:      17763  MISS:      54761
LLC RFO       ACCESS:       4853  HIT:       1275  MISS:       3578
LLC PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
LLC WRITEBACK ACCESS:      16484  HIT:      16394  MISS:         90
LLC PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
LLC AVERAGE MISS LATENCY: 163.1 cycles`

and here's the output from this PR running that same command:
`CPU 0 cumulative IPC: 1.17827 instructions: 50000002 cycles: 42435090
L1D TOTAL     ACCESS:   17052374  HIT:   16891333  MISS:     161041
L1D LOAD      ACCESS:    7294003  HIT:    7150506  MISS:     143497
L1D RFO       ACCESS:    9758371  HIT:    9740827  MISS:      17544
L1D PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
L1D WRITEBACK ACCESS:          0  HIT:          0  MISS:          0
L1D TRANSLATION ACCESS:          0  HIT:          0  MISS:          0
L1D PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
L1D AVERAGE MISS LATENCY: 90.6273 cycles
L1I TOTAL     ACCESS:    5238344  HIT:    4980158  MISS:     258186
L1I LOAD      ACCESS:    5238344  HIT:    4980158  MISS:     258186
L1I RFO       ACCESS:          0  HIT:          0  MISS:          0
L1I PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
L1I WRITEBACK ACCESS:          0  HIT:          0  MISS:          0
L1I TRANSLATION ACCESS:          0  HIT:          0  MISS:          0
L1I PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
L1I AVERAGE MISS LATENCY: 24.8702 cycles
L2C TOTAL     ACCESS:     475761  HIT:     198352  MISS:     277409
L2C LOAD      ACCESS:     401683  HIT:     155514  MISS:     246169
L2C RFO       ACCESS:      17543  HIT:       7539  MISS:      10004
L2C PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
L2C WRITEBACK ACCESS:      48234  HIT:      29605  MISS:      18629
L2C TRANSLATION ACCESS:       8301  HIT:       5694  MISS:       2607
L2C PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
L2C AVERAGE MISS LATENCY: 67.6785 cycles
ITLB TOTAL     ACCESS:    4474952  HIT:    4457422  MISS:      17530
ITLB LOAD      ACCESS:    4474952  HIT:    4457422  MISS:      17530
ITLB RFO       ACCESS:          0  HIT:          0  MISS:          0
ITLB PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
ITLB WRITEBACK ACCESS:          0  HIT:          0  MISS:          0
ITLB TRANSLATION ACCESS:          0  HIT:          0  MISS:          0
ITLB PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
ITLB AVERAGE MISS LATENCY: 4.1648 cycles
DTLB TOTAL     ACCESS:   10000305  HIT:    9856671  MISS:     143634
DTLB LOAD      ACCESS:    5851931  HIT:    5722626  MISS:     129305
DTLB RFO       ACCESS:    4148374  HIT:    4134045  MISS:      14329
DTLB PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
DTLB WRITEBACK ACCESS:          0  HIT:          0  MISS:          0
DTLB TRANSLATION ACCESS:          0  HIT:          0  MISS:          0
DTLB PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
DTLB AVERAGE MISS LATENCY: 6.56643 cycles
STLB TOTAL     ACCESS:     161164  HIT:     154234  MISS:       6930
STLB LOAD      ACCESS:     146835  HIT:     140725  MISS:       6110
STLB RFO       ACCESS:      14329  HIT:      13509  MISS:        820
STLB PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
STLB WRITEBACK ACCESS:          0  HIT:          0  MISS:          0
STLB TRANSLATION ACCESS:          0  HIT:          0  MISS:          0
STLB PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
STLB AVERAGE MISS LATENCY: 76.6063 cycles
PSCL5 TOTAL     ACCESS:       6930  HIT:       6930  MISS:          0
PSCL5 LOAD      ACCESS:       6110  HIT:       6110  MISS:          0
PSCL5 RFO       ACCESS:        820  HIT:        820  MISS:          0
PSCL5 PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
PSCL5 WRITEBACK ACCESS:          0  HIT:          0  MISS:          0
PSCL5 TRANSLATION ACCESS:          0  HIT:          0  MISS:          0
PSCL5 PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
PSCL5 AVERAGE MISS LATENCY: -nan cycles
PSCL4 TOTAL     ACCESS:       6930  HIT:       6930  MISS:          0
PSCL4 LOAD      ACCESS:       6110  HIT:       6110  MISS:          0
PSCL4 RFO       ACCESS:        820  HIT:        820  MISS:          0
PSCL4 PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
PSCL4 WRITEBACK ACCESS:          0  HIT:          0  MISS:          0
PSCL4 TRANSLATION ACCESS:          0  HIT:          0  MISS:          0
PSCL4 PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
PSCL4 AVERAGE MISS LATENCY: -nan cycles
PSCL3 TOTAL     ACCESS:       6930  HIT:       6930  MISS:          0
PSCL3 LOAD      ACCESS:       6110  HIT:       6110  MISS:          0
PSCL3 RFO       ACCESS:        820  HIT:        820  MISS:          0
PSCL3 PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
PSCL3 WRITEBACK ACCESS:          0  HIT:          0  MISS:          0
PSCL3 TRANSLATION ACCESS:          0  HIT:          0  MISS:          0
PSCL3 PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
PSCL3 AVERAGE MISS LATENCY: -nan cycles
PSCL2 TOTAL     ACCESS:       6930  HIT:       5506  MISS:       1424
PSCL2 LOAD      ACCESS:       6110  HIT:       5015  MISS:       1095
PSCL2 RFO       ACCESS:        820  HIT:        491  MISS:        329
PSCL2 PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
PSCL2 WRITEBACK ACCESS:          0  HIT:          0  MISS:          0
PSCL2 TRANSLATION ACCESS:          0  HIT:          0  MISS:          0
PSCL2 PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
PSCL2 AVERAGE MISS LATENCY: 0 cycles
LLC TOTAL     ACCESS:     293081  HIT:     196832  MISS:      96249
LLC LOAD      ACCESS:     246169  HIT:     159255  MISS:      86914
LLC RFO       ACCESS:      10004  HIT:       4613  MISS:       5391
LLC PREFETCH  ACCESS:          0  HIT:          0  MISS:          0
LLC WRITEBACK ACCESS:      34301  HIT:      32734  MISS:       1567
LLC TRANSLATION ACCESS:       2607  HIT:        230  MISS:       2377
LLC PREFETCH  REQUESTED:          0  ISSUED:          0  USEFUL:          0  USELESS:          0
LLC AVERAGE MISS LATENCY: 165.487 cycles`

The IPC of this PR is down about 11%, which is fine, but I'm more curious about what's going on with the L2C stats.  The translations can directly account for 2607 L2C misses, but it seems suuuuuper unlikely that would have a cascading effect that indirectly leads to 3.5x the total number of L2C misses.  Can you explain why we have 3.4x the load misses, 2.1x the RFO misses, and 83.9x the writeback misses?

It's also possible that I made a mistake in this test.  Can anyone else verify this, or offer an explanation of what's going on?  Thanks.

File: src/ptw.cc
Comment (ngober) [inline src/ptw.cc:94]: Rather than this if statement, you should rely on the underlying `MemoryRequestConsumer::add_rq()` to return -2 if the queue is full. This permits entries to be merged, which can be particularly important for high-order page table pages.

File: inc/ptw.h
Comment (ngober) [inline inc/ptw.h:26]: Only one inflight PTW seems unnecessarily restrictive. Given the potentially long latency of the walks, it's realistic that a few may be happening in parallel.

File: inc/ptw.h
Comment (ngober) [inline inc/ptw.h:29]: Consider including these as entries in the configuration file. This allows others to modify them easily as is appropriate for their research.

**Review by ngober — CHANGES_REQUESTED** [review]
I like this patch quite a lot. I think too much of it relies on the fact that PTW_MSHR_SIZE is 1, which doesn't have to be the case. If someone comes through and changes that value for their research, this will likely break. I want to continue talking about it, but I anticipate that this could be ready to merge before too long.

File: inc/memory_class.h
Comment (sethpugsley) [reply inc/memory_class.h:14]: IMO, we should be able to identify these translation loads, either through a separate type like this, or by adding a bit to the PACKET object.

File: src/main.cc
Comment (sethpugsley) [reply src/main.cc:658]: I agree that in the final version these should be either blocked off by an ifdef or commented out (but still present so it's obvious how to reenable it).

File: inc/ooo_cpu.h
Comment (sethpugsley) [reply inc/ooo_cpu.h:152]: https://download.01.org/perfmon/index/broadwell_server.html

I think this link also supports this idea.  Search this page for "PAGE_WALKER_LOADS."

File: inc/ptw.h
Comment (sethpugsley) [reply inc/ptw.h:26]: I think that building the inner workings of the page miss handler on top of the MSHR isn't quite the right thing to do at all.  I think that other than the producer/consumer interfaces between the STLB on one side and the caches on the other, the page table walker merits being its own thing that isn't just cobbled together from recycled parts.

Having said that, I think it's fine how it is for a first draft.  I wouldn't require this to be changed for the initial PR, and I think we could iterate on it in the long term.

Comment (ngober) [reply inc/ptw.h:26]: I'm not sure if I understand you. I think "MSHR" here is just a term to mean "in-progress page walks". It's not an MSHR in the same sense as the caches have.

Comment (sethpugsley) [reply inc/ptw.h:26]: I guess I agree that "miss status handling register" is not a bad name for this item.  Anyway, I still think this is something that could wait to be changed later.

File: inc/ptw.h
Comment (vishalgupta97) [reply inc/ptw.h:92]: I am reusing CACHE class as I am also tracking stats for PSCs (Reset/Print/Record stats function). If someone want to add replacement/prefetcher for PSCs, then it can also be done with the CACHE class.

If required for this PR, I will create a new class for PSC and add these three functions.

**Comment by vishalgupta97** [comment]
@sethpugsley: I added the dummy VA before the vmem class was implemented. I will convert my code to use get_pte_pa function.

The increase in total misses at L2 was due to a bug in PTW. Without PTW, STLB removes the last 12 bits from the physical address received from va_to_pa function, and then store the remaining value in the "data" variable before returning. With PTW, I was returning full physical address without removing the last 12 bits. I fixed that and now there is an increase of only 4% for that trace, due to translation misses.

File: inc/ptw.h
Comment (ngober) [reply inc/ptw.h:92]: Ah, now I see what you were going for here. I still think the class is too heavyweight for your purposes. You're instantiating 4 objects of type `CACHE` that never reference the bulk of the code in cache.cc. I think you should go with something simpler. If it were me, I would make a `std::vector<BLOCK>` and track the stats locally. If you want to make a simple class for it, that works just as well.

If you're concerned about being able to use the `print_roi_stats()` family of functions, I would make a `PageTableWalker::print_stats()` that you can call at the end of the simulation.

**Comment by vishalgupta97** [comment]
Added code to handle multiple outstanding translation at PTW. Currently, the code has some issues due to merging at L1D. I am fixing that.

1. I am having some issue in migrating the PTW constants to config.sh file. Where should the PTW constants be added ?
2. Currently I am not using get_pte_pa from vmem.cc because there are IPC traces in which upper bits are all zero, so the lookup_tag is zero for Level 0 as well as Level 1.

File: src/ptw.cc
Comment (ngober) [inline src/ptw.cc:52]: I think you should rely on the lower level to know whether its RQ is full or not. The lower level will return -2 if the insertion failed. Then, you don't need a cast to `CACHE*`, or even the check here. Also, consider using the function `delay_queue::has_ready()`, since the RQ is a delay queue.

**Review by ngober — CHANGES_REQUESTED** [review]
This latest revision is a big step forward towards being merge-worthy. I think there are still things to clean up, but the core of it is there.

I still disagree with the decision to use the `CACHE` class for the PSCLs. I think the way that it is being used is clunky and tacked-on. I also have some style disagreements, such as the idiom

    while (X) {
        if (Y) {
            ...
        } else {
            break;
        }
    }

But, I don't think style changes make this un-mergeable. Style can always be tweaked later. This is all very close and I am excited to be able to merge it soon.

**Comment by ngober** [comment]
I want to respond to these questions inline:

>     1. I am having some issue in migrating the PTW constants to config.sh file. Where should the PTW constants be added ?

A short-term solution will be to add them as entries in the `const_names` dict, so that they end up in inc/champsim_constants.h. However, this probably isn't compatible with #124. In this case, a better long-term solution would be to pass the PTW constants into the constructor, then pass a pointer to the PTW into the `O3_CPU` constructor. See the other pull request for an example of how I did this with the caches.

>     2. Currently I am not using get_pte_pa from vmem.cc because there are IPC traces in which upper bits are all zero, so the lookup_tag is zero for Level 0 as well as Level 1.

I wouldn't worry too much about the IPC traces. They've had some issues with the address space being weird, and they're probably atypical. The problem is worth addressing, but it shouldn't stop this patch from moving forward.

File: src/cache.cc
Comment (vishalgupta97) [reply src/cache.cc:52]: I have created MSHR as a list of PACKET, same as how its done in cache. I am reusing the code for MSHR at caches for PTW. Maybe these functions can be moved to util.h or I can create a copy in ptw.h

File: inc/ptw.h
Comment (vishalgupta97) [reply inc/ptw.h:81]: Yes. I will remove this.

File: src/ptw.cc
Comment (vishalgupta97) [reply src/ptw.cc:50]: Yes. This is a bug. MSHR should be checked in each iteration of while loop.

File: src/cache.cc
Comment (ngober) [reply src/cache.cc:52]: I think the best place would be either util.h or block.h, since these structures are mostly concerned with PACKET objects.

**Comment by vishalgupta97** [comment]
I have implemented all the changes requested. The default values for PSCs, PTW RQ, MSHR can be updated based on the real system.

**Comment by sethpugsley** [comment]
I finally got around to testing out the latest version of this PR today.  Sorry it took so long.  I ran it for 10M warmup & 10M sim instructions on all the IPC-1 traces, and I didn't find any problems with it.  Geomean IPC change was -1.5%, largest positive IPC change was +1.1%, and largest negative IPC change was -5.8%.  I didn't encounter any crashes or other bugs that I know of.

@ngober, do you think that it's ready to go?  As far as I can tell, it's ready, but I'll leave it to you to give final approval and actually perform the merge.  Thanks for all your hard work, @vishalgupta97!

**Review by ngober — APPROVED** [review]
I like these changes. I have some style comments, and `VirtualMemory::get_pte_pa()` is still an orphan function, but I think this patch is at a point where it can be merged. Since Seth has signed off, I'm going to go ahead and merge it immediately. Thank you for your contribution, Vishal!

**Pull request merged**
